### PR TITLE
[Caching] Create clang importer from cc1 args directly

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
When caching build is enabled, teach dependency scanner to report command-lines with `-direct-clang-cc1-module-build` so the later compilation can instantiate clang importer with cc1 args directly. This avoids running clang driver code, which might involve file system lookups, which are the file deps that are not captured and might result in different compilation mode.

rdar://119275464

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
Currently, only enable the new code path when caching is enabled since it is quite a big behavior change. Caching build requires the change since its restricted file system will definitely alter the behavior for clang importer. For normal explicit module build, this will still help with less lookup during the actual compilation, but we can enable that later.

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
Rebase to resolve conflict. After some extensive testing, I feel pretty confident to turn this on for caching only. Ready to be reviewed.

@artemcm @DougGregor @benlangmuir @beccadax Since this is a pretty big behavior change, do you have any concern turn on this for all explicit module build? Will run a separate qualification for that.

**Comment by artemcm** [comment]
> Rebase to resolve conflict. After some extensive testing, I feel pretty confident to turn this on for caching only. Ready to be reviewed.
> 
> @artemcm @DougGregor @benlangmuir @beccadax Since this is a pretty big behavior change, do you have any concern turn on this for all explicit module build? Will run a separate qualification for that.

Given sufficient qualification I don't have a concern with enabling this for all Explicit Module Builds. But it would be good to have that enablement be a separate change after this one that we hold off merging until we're able to get enough confidence/validation, or land behind a flag temporarily which we can use to do testing with.

**Comment by benlangmuir** [comment]
> Since this is a pretty big behavior change, do you have any concern turn on this for all explicit module build? Will run a separate qualification for that.

Doing this for all explicit builds seems best to me.  Not only this reduces the differences in behaviour for caching, but it also means that clang modules used by swift will behave more like clang modules used by clang -- clang's dependency scanner has been using -cc1 for explicit module builds for a while now.

> But it would be good to have that enablement be a separate change after this one that we hold off merging until we're able to get enough confidence/validation, or land behind a flag temporarily which we can use to do testing with.

Flag seems like a good idea to me to make it easier to control until we're ready to flip the default for non-caching.

**Comment by cachemeifyoucan** [comment]
Add a flag to turn on this new behavior for non-caching build. Add some nice to have swift-driver support here: https://github.com/apple/swift-driver/pull/1550

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test linux platform

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
Rebase to resolve merge conflict. Still waiting for review :)

File: lib/ClangImporter/ClangImporter.cpp
Comment (benlangmuir) [inline lib/ClangImporter/ClangImporter.cpp:3977]: Maybe use `instance.clearImplicitModuleBuildOptions()`?

File: lib/ClangImporter/ClangImporter.cpp
Comment (benlangmuir) [inline lib/ClangImporter/ClangImporter.cpp:3980]: Can we switch this to use `tooling::dependencies::configureInvocationForCaching` by passing in the include-tree root ID to this function (optionally, if caching is enabled)?

File: lib/ClangImporter/ClangImporter.cpp
Comment (benlangmuir) [inline lib/ClangImporter/ClangImporter.cpp:1111]: Don't we already clear this from the direct cc1 args in `getSwiftExplicitModuleDirectCC1Args`?

File: lib/ClangImporter/ClangImporter.cpp
Comment (cachemeifyoucan) [reply lib/ClangImporter/ClangImporter.cpp:3980]: That is the eventual goal. But note following for current setup. This is creating a ClangImporter that is not using include-tree even when caching is enabled, but a regular translation unit that imports modules that are built with include tree. It needs a different set of options from an include tree build, thus it has a customized option clearing function and may need some other tweaking.
In the future, we should run a dependency scanning of the clang importer instance then let clang scanner return the arguments.

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
ping~

File: lib/ClangImporter/ClangImporter.cpp
Comment (cachemeifyoucan) [reply lib/ClangImporter/ClangImporter.cpp:3979]: I think that makes sense if we have depscanning. Right now, it is matching the behavior in
`ClangImporterOptions::getReducedExtraArgsForSwiftModuleDependency`, just in a more elegant way.

**Review by benlangmuir — APPROVED** [review]
LGTM, but probably deserves review from @artemcm as well

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@artemcm Ping. Most of the PR should have no changes to non-caching build unless the new option is explicated used. The only behavior changing part is separated out as 9f736811f0a6bdc87b5af81dce00929a48fb0673. Let me know if this is good to merge as I have lots of other patches that can cause conflict with this one.

**Review by artemcm — APPROVED** [review]
This looks good. Thank you @cachemeifyoucan

**Comment by drodriguez** [comment]
I am trying to find the reason for an internal failure in a Linux system. For it to work, we need to provide Clang with `--gcc-toolchain=/foo/bar` to point to the GCC toolchain in a non-standard location, so we normally use `-Xcc --gcc-toolchain=/foo/bar`. It seems that this PR might forward that parameter (which is intended for the Clang driver) to cc1, failing many of the tests with `unknown argument: '--gcc-toolchain=/foo/bar' … error: clang importer creation failed`.

Is there a way of still using the Clang driver or is that completely gone? Is there a good place to filter then non-cc1 arguments somehow? I am not sure that removing `--gcc-toolchain` from the command lines will work at all, since it expands to many search paths and options (and I can imagine some other similar options that the Clang driver expands to many options in cc1).

**Comment by cachemeifyoucan** [comment]
> I am trying to find the reason for an internal failure in a Linux system. For it to work, we need to provide Clang with `--gcc-toolchain=/foo/bar` to point to the GCC toolchain in a non-standard location, so we normally use `-Xcc --gcc-toolchain=/foo/bar`. It seems that this PR might forward that parameter (which is intended for the Clang driver) to cc1, failing many of the tests with `unknown argument: '--gcc-toolchain=/foo/bar' … error: clang importer creation failed`.
> 
> Is there a way of still using the Clang driver or is that completely gone? Is there a good place to filter then non-cc1 arguments somehow? I am not sure that removing `--gcc-toolchain` from the command lines will work at all, since it expands to many search paths and options (and I can imagine some other similar options that the Clang driver expands to many options in cc1).

The new option is supposed to be only used for explicit module build, which requires dependency scanning. The `-Xcc` command passed to the swift-driver (or swiftc) is still clang driver flag and that has not changed. There has always been a special mode (mostly for building PCM) that changes the meaning of `-Xcc` flag to swift-frontend to be clang cc1 flag, from which you have to provide all the arguments to initialize clang importer. This patch just add a new option to ask swift dependency scanner to return options to initialize all swift-frontend directly with cc1 options.

There is a follow-up PR in swift-driver to tell it not to forward `-Xcc` option from the command-line to swift-frontend anymore. https://github.com/apple/swift-driver/pull/1550. It doesn't need to be forwarded because swift dependency scanner has already seen the driver flag and created the list of cc1 flag that is corresponding driver options and swift-frontend does not need to run clang driver again.

**Pull request merged**
